### ISSUE TITLE
Add Mismatches matcher to test Matcher's mistatches

### ISSUE THREAD

**Issue by victornoel** [llorllale/cactoos-matchers#106]
There exists a `Matcher` called  `Matches` to test that a `Matcher` properly matches.

We need a complementary `Matcher` named `Mismatches` to test that a `Matcher` properly does not match, including checking the mismatch message.

For example, in order to do #85, I must remove the usage of `ExpectedException` and most of the place where it is used, the intent is actually about testing that the `Matcher` properly mismatch with the correct message.

Scope:
1. introduce new `Matcher` `Mismatches` with constructors:
  - `Mismatches(T arg, String expected, String actual)`
  - `Mismatches(T arg, Text expected, Text actual)`
  - `Mismatches(T arg, String message)`
  - `Mismatches(T arg, Text message)`

Its usage would be for example:
```java
        new Assertion<>(
            "Can't describe a mismatch",
            new HasLines("Tom", "Mike"),
            new Mismatches<>(
                String.format("Tom%nJohn%n"),
                "Lines are <[Tom, Mike]>",
                "<[Tom, John]>"
            )
        ).affirm();
```
or 
```java
        new Assertion<>(
            "Can't describe a mismatch",
            new HasLines("Tom", "Mike"),
            new Mismatches<>(
                String.format("Tom%nJohn%n"),
                "Expected: Lines are <[Tom, Mike]>\n but was: <[Tom, John]>"
            )
        ).affirm();
```
- At least one arbitrary test should be converted to its use as an example (`HasLines` above is a perfect candidate)

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by victornoel** [comment]
@llorllale ping, it's blocking #85 :)

**Comment by llorllale** [comment]
@victornoel this won't work with `Throws`. How about making the first arg a `Scalar<T>`?

**Comment by victornoel** [comment]
@llorllale hmm, if the `Matcher` under test is a `Throws`, then it means that in the signature above, the first argument `arg` would just been instantiated as `Scalar<U>` for some `U`, no?

Actually, I'm not exactly sure why `Assertion` is not simply the same, taking a `T` (which in the case of `Throws` matcher is instantiated as a `Scalar<U>` too)... That would solve this strange double-constructor in `Assertion`, no?

Well I suppose this may have already been tried?

**Comment by llorllale** [comment]
@victornoel 

> I'm not exactly sure why Assertion is not simply the same, taking a T

**TLDR:** `Assertion` should be capable of asserting values *and exceptions*. It must accept a `Scalar<T>` so that it can catch any exception thrown and use the provided matcher (which *may* be `Throws`), to test the exception itself.

**Longer version:**

Because `Throws` requires the operation to be executed and allowed to actually throw the exception. The operation is executed by `Assertion` via `Scalar.value()`. Remember, ["Assertions are always performed on *operations*, as opposed to actual values, in order to also enable asserting on errors"](https://github.com/llorllale/cactoos-matchers/blob/6af8e10f975742e20d7bb4807c60697327625d19/src/main/java/org/llorllale/cactoos/matchers/Assertion.java#L40). If `Assertion` were to simply accept `T` as an argument, then passing in a `Scalar<T>` would be treated as a value to be compared, not an operation to execute, therefore making it impossible to assert that a given operation throws an exception.

On to `Mismatches`:
The above has ramifications for both `Matches` and this new `Mismatches` you're proposing. Let's imagine we want to assert that `Throws` does *not* match on `errorIfInputIs3` when the input is different than 3. What happens if `Mismatches` does *not* accept a `Scalar<T>`: 

```java
        new Assertion<>(
            "Can't describe a mismatch",
            () -> new Throws(Exception.class),
            new Mismatches<>(
               errorIfInputIs3(0)
            )
        ).affirm();
```

Answer: the exception will not be caught by `Assertion`, but rather the surrounding context. You'd be forced to use either `org.junit.Test#expected`, `ExpectedException` rule, try/catch, or anything else other than `Assertion`.

**Comment by victornoel** [comment]
@llorllale I think you are confusing the type argument `T` and its actual instance which can itself be a `Scalar<X>`. I understand what you say above above the requirements for `Assertion` and agree with them.

I disagree on the technical solution.

See the following code I just wrote to demonstrate my point:
```java
public final class Assertion2<T> {

	private final T arg;
	private final Matcher<T> matcher;

	public Assertion2(T arg, Matcher<T> matcher) {
		this.arg = arg;
		this.matcher = matcher;
	}

	public void affirm() {
		if (!matcher.matches(arg)) {
			throw new AssertionError();
		}
	}
}

public final class Tests {

	@Test
	public void testText() {
		new Assertion2<>(new TextOf("a"), new TextIs("a")).affirm();
	}

	@Test
	public void testThrows() {
		new Assertion2<Scalar<Integer>>(
			() -> {
				if (true) {
					throw new IllegalStateException("a");
				}
				return 1;
			},
			new Throws<>("a", IllegalStateException.class)
		).affirm();
	}
}
```

Do you see any problem with this? It answers the requirements you expressed and doesn't need that `Assertion` take a `Scalar`. It takes the types `T` that is required by the `Matcher`. When the `Matcher` requires a `T` that is a `Scalar`, then an instance of `Assertion` will gladly take a `Scalar` as the code example above shows.

**Comment by llorllale** [comment]
@victornoel 

> Do you see any problem with this? 

Incompatibility between the generic types of `Assertion` and `Throws`. And `Throws` requires *something* it can execute (eg. a `Scalar`) in order to trigger the exception for its test.

**Comment by victornoel** [comment]
> Incompatibility between the generic types of Assertion and Throws. And Throws requires something it can execute (eg. a Scalar) in order to trigger the exception for its test.

What do you mean incompatibility? They are perfectly compatible:
- `Assertion` takes a `Matcher<T>` whatever is the instantiation of `T`.
- `Throws` is a `Matcher<T>` where `T` is instantiated with `Scalar<X>` for any `X` (generics are relative to their declaration, so the `X` here is the `T` in `Throws<T>`, which is NOT the `T` in `Assertion<T>` and that's perfectly ok).

Have you realised that the code above is compiling and that the tests both passes, and that this demonstrate the point I am making?

**Comment by victornoel** [comment]
@llorllale let me add something else: in the example code above, in the test `testThrows()`, you can see that:
- the object `Assertion2` takes an object `Scalar` as you expect it to take so that it is executable as you highlight
- it is possible because `Throws` is a `Matcher<Scalar<T>>`, so it can only matches against instances of `Scalar<T>`. Here `T` is instantiated as `Integer`, but it is not really used. Actually `Throws` could be a `Matcher<Proc>` and `Assertion2` wouldn't have to be changed any less.
- and thus an exception thrown in it is caught by `Throws` and not before as you would expect

**Comment by llorllale** [comment]
@victornoel you're right. You pointed out my erroneous assumption:

> generics are relative to their declaration, so the X here is the T in Throws<T>, which is NOT the T in Assertion2<T> 

so my understanding was that `Throws.matches()` would be passed a `Scalar<T>` where `T` would be another `Scalar`. This boggles my mind. Any references explaining this would be appreciated.

This would make using `Assertion` a bit more intuitive (no more unnecessary scalars when asserting on *values*).

Can you open a ticket to refactor `Assertion`?

**Comment by llorllale** [comment]
@0crat in

let's do `Mismatches`

**Comment by 0crat** [comment]
> @0crat in let's do `Mismatches` ([here](https://github.com/llorllale/cactoos-matchers/issues/106#issuecomment-483660117))

@llorllale Argument "let's" doesn't match regex `[A-Z]{2,3}`

**Comment by llorllale** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/llorllale/cactoos-matchers/issues/106#issuecomment-483660787))

@llorllale Job [#106](https://github.com/llorllale/cactoos-matchers/issues/106) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://www.zerocracy.com/policy.html#29): +15 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by victornoel** [comment]
@llorllale I believe #88 is the perfect place for this to happen, WDYT? If yes, I will add a comment there to explain the solution we described here.

As for a pointer on type parameters, I'm not exactly sure where to look for it. It's "basic" in the sense that that's how they work: there are type parameters when a generic is declared, and type argument when a generic is instantiated (potentially to declare another generic).
This particular aspect can sometimes be quite implicit (especially if you haven't wrote a compiler before ^^), it's the same thing that often confuse beginners between method parameters and method arguments.

**Comment by llorllale** [comment]
@victornoel 

>  I believe #88 is the perfect place for this to happen, WDYT? If yes, I will add a comment there to explain the solution we described here.

Sure, go ahead. Link back to this conversation.

> As for a pointer on type parameters, I'm not exactly sure where to look for it. It's "basic" in the sense that that's how they work: there are type parameters when a generic is declared, and type argument when a generic is instantiated (potentially to declare another generic).
This particular aspect can sometimes be quite implicit (especially if you haven't wrote a compiler before ^^), it's the same thing that often confuse beginners between method parameters and method arguments.

What's confusing is that this *thing* I'm passing through `Assertion` and on to `Throws` has a different "appearance" based on the "context" (Assertion vs Throws).

**Comment by 0crat** [comment]
The job [#106](https://github.com/llorllale/cactoos-matchers/issues/106) assigned to @longstone[/z](https://www.0crat.com/u/longstone), here is [why](https://www.0crat.com/footprint/C63314D6Z/383ffb1e-36f7-452a-a009-b141c29afbdf); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by victornoel** [comment]
@longstone are you going to tackle that? This is needed for me to solve other issues :)

**Comment by victornoel** [comment]
@llorllale I think @longstone is not going to take care of this issue, can it be re-assigned by 0crat because I'm waiting for it to work on #114

**Comment by llorllale** [comment]
@0crat refuse

**Comment by llorllale** [comment]
@0crat assign @victornoel

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/llorllale/cactoos-matchers/issues/106#issuecomment-502477650))

@llorllale The user @longstone[/z](https://www.0crat.com/u/longstone) resigned from [#106](https://github.com/llorllale/cactoos-matchers/issues/106), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat assign @victornoel ([here](https://github.com/llorllale/cactoos-matchers/issues/106#issuecomment-502477672))

@llorllale The job [#106](https://github.com/llorllale/cactoos-matchers/issues/106) assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/2ad25c6d-ea01-4205-89b2-6bd3feb282d3); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by 0crat** [comment]
It is strongly discouraged to assign jobs to their creators, see [§19](http://www.zerocracy.com/policy.html#19): -15 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by victornoel** [comment]
@0crat wait for #131 to be merged

**Comment by 0crat** [comment]
> @0crat wait for #131 to be merged ([here](https://github.com/llorllale/cactoos-matchers/issues/106#issuecomment-504676686))

@victornoel The impediment for [#106](https://github.com/llorllale/cactoos-matchers/issues/106) was registered successfully by @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0pdd** [comment]
@victornoel 3 puzzles [#136](https://github.com/llorllale/cactoos-matchers/issues/136), [#137](https://github.com/llorllale/cactoos-matchers/issues/137), [#138](https://github.com/llorllale/cactoos-matchers/issues/138) are still not solved.

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#106](https://github.com/llorllale/cactoos-matchers/issues/106) is now out of scope

**Comment by sereshqua** [comment]
@victornoel please make sure that problem reporter (you) and problem solver (you) would be two different persons next time :)

**Comment by victornoel** [comment]
@sereshqua yes, I agree

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +30 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

**Comment by 0pdd** [comment]
@victornoel 3 puzzles [#137](https://github.com/llorllale/cactoos-matchers/issues/137), [#138](https://github.com/llorllale/cactoos-matchers/issues/138), [#164](https://github.com/llorllale/cactoos-matchers/issues/164) are still not solved; solved: [#136](https://github.com/llorllale/cactoos-matchers/issues/136).

**Comment by 0pdd** [comment]
@victornoel 2 puzzles [#137](https://github.com/llorllale/cactoos-matchers/issues/137), [#164](https://github.com/llorllale/cactoos-matchers/issues/164) are still not solved; solved: [#136](https://github.com/llorllale/cactoos-matchers/issues/136), [#138](https://github.com/llorllale/cactoos-matchers/issues/138).

**Comment by 0pdd** [comment]
@victornoel the puzzle [#164](https://github.com/llorllale/cactoos-matchers/issues/164) is still not solved; solved: [#136](https://github.com/llorllale/cactoos-matchers/issues/136), [#137](https://github.com/llorllale/cactoos-matchers/issues/137), [#138](https://github.com/llorllale/cactoos-matchers/issues/138).
